src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java

```diff
@@ -486,46 +516,38 @@ private boolean checkSynchronized(final DetailAST synchronizedAst, boolean useBr
      * </pre>
      *
      * @param currentCase AST of the case that falls through to the next case.
-     * @param nextCase AST of the next case.
      * @return True if a relief comment was found
      */
-    private boolean hasFallThroughComment(DetailAST currentCase, DetailAST nextCase) {
-        boolean allThroughComment = false;
-        final int endLineNo = nextCase.getLineNo();
-
-        if (matchesComment(reliefPattern, endLineNo)) {
-            allThroughComment = true;
+    private boolean hasFallThroughComment(DetailAST currentCase) {
+        final DetailAST nextCase = currentCase.getNextSibling();
+        final DetailAST ast;
+        if (nextCase.getType() == TokenTypes.CASE_GROUP) {
+            ast = nextCase.getFirstChild();
         }
         else {
-            final int startLineNo = currentCase.getLineNo();
-            for (int i = endLineNo - 2; i > startLineNo - 1; i--) {
-                final int[] line = getLineCodePoints(i);
-                if (!CodePointUtil.isBlank(line)) {
-                    allThroughComment = matchesComment(reliefPattern, i + 1);
-                    break;
-                }
-            }
+            ast = currentCase;
         }
-        return allThroughComment;
+        return isReliefComment(ast);
     }
 
     /**
-     * Does a regular expression match on the given line and checks that a
-     * possible match is within a comment.
+     * Check if there is any fall through comment.
      *
-     * @param pattern The regular expression pattern to use.
-     * @param lineNo The line number in the file.
-     * @return True if a match was found inside a comment.
+     * @param ast ast to check
+     * @return true if relief comment found
      */
-    // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
-    @SuppressWarnings("deprecation")
-    private boolean matchesComment(Pattern pattern, int lineNo) {
-        final String line = getLine(lineNo - 1);
-
-        final Matcher matcher = pattern.matcher(line);
-        return matcher.find()
-                && getFileContents().hasIntersectionWithComment(
-                        lineNo, matcher.start(), lineNo, matcher.end());
+    private boolean isReliefComment(DetailAST ast) {
+        boolean ans = false;
+        DetailAST currentAst = ast;
+        while (currentAst != null) {
+            if (currentAst.getNextSibling() != null
+                    && currentAst.getNextSibling().getType() != TokenTypes.SINGLE_LINE_COMMENT
+                    && currentAst.getNextSibling().getType() != TokenTypes.BLOCK_COMMENT_BEGIN) {
+                ans = reliefPattern.matcher(currentAst.getFirstChild().getText()).find();
```

Comment:
The logic here is confusing to me. Why are we checking for any token besides `COMMENT_CONTENT` or at least a comment delimiter? Why can't we look for comment delimiters, then extract the content , then try to match the regexp?

Reply:
I have not tried to write the logic only base of `COMMENT_CONTENT`. 
But the loop is there in method becuase for code 
```
        switch (i) {
            case 0:
                i++;
                /* comment */ /* comment */
                /* fall thru */
        }
```

in this we have to check all the comments whether there is fall through or not In my opinion the current logic is better for iteration. either to first get the first child(`COMMENT_CONTENT`) of single or Block comment and put extra condition to check wether it is null or not

Reply:
This still seems like too much work to me. How is what we are doing here any different than the following:
1. Get next ast that is NOT a comment type
2. Check that ast's previous sibling's first child

This saves us a lot of extra checking (only one regexp match per method call) and extra code. In fact, we can probably just use the new method you've written, `getNextNonCommentAst`.

Reply:
> This still seems like too much work to me. How is what we are doing here any different than the following:

I totally agree The main work is still pending which will be resolved at https://github.com/checkstyle/checkstyle/issues/13545